### ISSUE TITLE
[jvm] Records, and handling various Java SDKs

### ISSUE THREAD

**Issue by EliteMasterEric** [HaxeFoundation/haxe#11016]
For one of my projects, I encountered an issue where `java.lang.Record` is not defined properly. I believe this can be resolved using externs, however it raises the question of how future Java versions should be handled.

Importantly, certain applications are not compatible with certain versions of Java, so if Haxe sticks to Java 8, it can't properly interact with Records (introduced in JDK 14), but if Haxe updates to Java 14, it will start generating `.class` files with a major version of `58`, meaning applications expecting a JAR compatible with Java 8 want `.class` files with a major version of `52`, and can't process classes generated with the newer JDK.

Haxe may need additional options to allow the user to define what JDK level they are compiling their application for, and enable the use of certain features accordingly.

**Comment by Simn** [comment]
Ideally this is entirely isolated by what https://github.com/HaxeFoundation/hxjava provides. IIRC @kevinresol found a good way to switch between different Java versions, so maybe he has some insight.

**Comment by EliteMasterEric** [comment]
I believe in the meantime I will attempt to switch to the `--java` target.

@kevinresol Do you have any insight into how to build `.class` files for newer JDK versions?

**Comment by kevinresol** [comment]
Hmm I can't recall anymore unfortunately. But I remember I didn't do much work besides generating something with the hxjava tool. Maybe this one: https://github.com/kevinresol/hxjava/commit/0ec5c0fb3b4651f0467e19191f2cda5589676681

**Comment by EliteMasterEric** [comment]
Over the course of developing my project on-and-off, I've concluded that the feature I'm looking for is easily available to the `--java` target, via using the `--c-arg` compiler argument to set the source and target versions. The issue is with the JVM, which always produces classes matching class version 52 (corresponding with Java 8).

Thus, the issue is not with `hxjava`, but rather with *EDIT* Found it, it's this line: https://github.com/HaxeFoundation/haxe/blob/b24ae6839db69a6d0b5df144ef1344010b074822/src/generators/jvm/jvmClass.ml#L218

I will adjust the issue name to clarify this.

**Comment by Simn** [comment]
We can add an option to configure this, but I don't really understand how that improves anything. My understanding is that if you want to deal with records, Haxe would have to support the [Record attribute](https://docs.oracle.com/en/java/javase/14/docs/specs/records-jvms.html#jvms-4.7.30) in some way.

**Comment by EliteMasterEric** [comment]
For the record, the statement "Importantly, certain applications are not compatible with certain versions of Java" was wrong. The issue I was experiencing was actually #11023. As long as your class loader is compliant with Java's standards, older JARs work with newer applications just fine.

Therefore, I do not see any use case for defining a target class version right now.

As for records, I made this extern:
```haxe
package java.lang;

/**
 * Records are a new kind of type declaration introduced in Java 14.
 * They are a restricted form of class ideal for immutable data carriers.
 */
@:dce
extern abstract class Record
{
  @:protected function new();
  public abstract function equals(obj:Dynamic):Bool;
  public abstract function hashCode():Int;
  public abstract function toString():String;
}
```

Then, I was able to create `extern class extends Record`, manually define corresponding constructor and getter functions and I haven't had any issues, thus official language support for the Record type isn't really necessary.
